File: LinkDemo-ObjC/LinkDemo-ObjC/ViewControllers/LayerSession/PlaidLayerSessionViewController.m
//
//  PlaidLayerSessionViewController.m
//  LinkDemo-ObjC
//
//  Copyright © 2026 Plaid Inc. All rights reserved.
//

#import "PlaidLayerSessionViewController.h"
#import "LinkKitObjectiveC.h"
@import LinkKitObjC;

typedef NS_ENUM(NSInteger, LoadingState) {
    LoadingStateLoading,
    LoadingStateReady,
    LoadingStateError
};

@interface PlaidLayerSessionViewController ()

@property (nonatomic, copy) NSString *linkToken;
@property (nonatomic, assign) LoadingState state;
@property (nonatomic, copy, nullable) NSString *errorMessage;
@property (nonatomic, strong, nullable) id<PLKPlaidLayerSession> layerSession;

@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *versionLabel;
@property (nonatomic, strong) UILabel *errorLabel;
@property (nonatomic, strong) UITextField *phoneNumberTextField;
@property (nonatomic, strong) UITextField *dobTextField;
@property (nonatomic, strong) UIButton *submitButton;
@property (nonatomic, strong) UIButton *launchButton;
@property (nonatomic, strong) UIActivityIndicatorView *activityIndicator;

@end

@implementation PlaidLayerSessionViewController

- (instancetype)initWithLinkToken:(NSString *)linkToken {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        _linkToken = [linkToken copy];
        _state = LoadingStateLoading;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"Plaid Layer Session";
    self.view.backgroundColor = [UIColor systemBackgroundColor];
    
    [self setupUI];
    [self createSession];
}

- (void)setupUI {
    UIScrollView *scrollView = [[UIScrollView alloc] init];
    scrollView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:scrollView];
    
    UIStackView *stackView = [[UIStackView alloc] init];
    stackView.axis = UILayoutConstraintAxisVertical;
    stackView.spacing = 16;
    stackView.translatesAutoresizingMaskIntoConstraints = NO;
    [scrollView addSubview:stackView];

    self.titleLabel = [[UILabel alloc] init];
    self.titleLabel.text = @"Plaid Layer Session Example";
    self.titleLabel.font = [UIFont boldSystemFontOfSize:20];
    self.titleLabel.textAlignment = NSTextAlignmentCenter;
    self.titleLabel.numberOfLines = 0;
    [stackView addArrangedSubview:self.titleLabel];

    self.versionLabel = [[UILabel alloc] init];
    NSString *version = [PLKPlaid sdkVersion];
    self.versionLabel.text = [NSString stringWithFormat:@"LinkKit v%@", version];
    self.versionLabel.font = [UIFont systemFontOfSize:14];
    self.versionLabel.textColor = [UIColor secondaryLabelColor];
    self.versionLabel.textAlignment = NSTextAlignmentCenter;
    [stackView addArrangedSubview:self.versionLabel];

    self.errorLabel = [[UILabel alloc] init];
    self.errorLabel.font = [UIFont systemFontOfSize:14];
    self.errorLabel.textColor = [UIColor systemRedColor];
    self.errorLabel.textAlignment = NSTextAlignmentCenter;
    self.errorLabel.numberOfLines = 0;
    self.errorLabel.hidden = YES;
    [stackView addArrangedSubview:self.errorLabel];

    UILabel *phoneLabel = [[UILabel alloc] init];
    phoneLabel.text = @"User Phone Number";
    phoneLabel.font = [UIFont systemFontOfSize:14];
    phoneLabel.textColor = [UIColor secondaryLabelColor];
    [stackView addArrangedSubview:phoneLabel];
    
    self.phoneNumberTextField = [[UITextField alloc] init];
    self.phoneNumberTextField.placeholder = @"+1 415-555-0011";
    self.phoneNumberTextField.text = @"+1 415-555-0011";
    self.phoneNumberTextField.keyboardType = UIKeyboardTypePhonePad;
    self.phoneNumberTextField.borderStyle = UITextBorderStyleRoundedRect;
    [stackView addArrangedSubview:self.phoneNumberTextField];

    UILabel *dobLabel = [[UILabel alloc] init];
    dobLabel.text = @"User Date of Birth";
    dobLabel.font = [UIFont systemFontOfSize:14];
    dobLabel.textColor = [UIColor secondaryLabelColor];
    [stackView addArrangedSubview:dobLabel];
    
    self.dobTextField = [[UITextField alloc] init];
    self.dobTextField.placeholder = @"YYYY-MM-DD";
    self.dobTextField.text = @"1975-01-18";
    self.dobTextField.keyboardType = UIKeyboardTypeNumbersAndPunctuation;
    self.dobTextField.borderStyle = UITextBorderStyleRoundedRect;
    [stackView addArrangedSubview:self.dobTextField];

    self.submitButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.submitButton setTitle:@"Submit User Data" forState:UIControlStateNormal];
    self.submitButton.titleLabel.font = [UIFont boldSystemFontOfSize:16];
    [self.submitButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.submitButton.backgroundColor = [UIColor systemBlueColor];
    self.submitButton.layer.cornerRadius = 10;
    [self.submitButton addTarget:self action:@selector(submitButtonTapped) forControlEvents:UIControlEventTouchUpInside];
    [stackView addArrangedSubview:self.submitButton];

    UIView *spacer = [[UIView alloc] init];
    [spacer setContentHuggingPriority:UILayoutPriorityDefaultLow forAxis:UILayoutConstraintAxisVertical];
    [stackView addArrangedSubview:spacer];

    self.launchButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.launchButton setTitle:@"Launch Layer" forState:UIControlStateNormal];
    self.launchButton.titleLabel.font = [UIFont boldSystemFontOfSize:16];
    [self.launchButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.launchButton.backgroundColor = [UIColor systemBlueColor];
    self.launchButton.layer.cornerRadius = 10;
    [self.launchButton addTarget:self action:@selector(launchButtonTapped) forControlEvents:UIControlEventTouchUpInside];
    [stackView addArrangedSubview:self.launchButton];

    self.activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleMedium];
    self.activityIndicator.color = [UIColor whiteColor];
    self.activityIndicator.hidesWhenStopped = YES;
    self.activityIndicator.translatesAutoresizingMaskIntoConstraints = NO;
    [self.launchButton addSubview:self.activityIndicator];

    [NSLayoutConstraint activateConstraints:@[
        [scrollView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [scrollView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [scrollView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [scrollView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],

        [stackView.topAnchor constraintEqualToAnchor:scrollView.topAnchor constant:16],
        [stackView.leadingAnchor constraintEqualToAnchor:scrollView.leadingAnchor constant:16],
        [stackView.trailingAnchor constraintEqualToAnchor:scrollView.trailingAnchor constant:-16],
        [stackView.bottomAnchor constraintEqualToAnchor:scrollView.bottomAnchor constant:-16],
        [stackView.widthAnchor constraintEqualToAnchor:scrollView.widthAnchor constant:-32],

        [self.submitButton.heightAnchor constraintEqualToConstant:56],
        [self.launchButton.heightAnchor constraintEqualToConstant:56],
        [self.activityIndicator.leadingAnchor constraintEqualToAnchor:self.launchButton.leadingAnchor constant:16],
        [self.activityIndicator.centerYAnchor constraintEqualToAnchor:self.launchButton.centerYAnchor]
    ]];
    
    [self updateUI];
}

- (void)createSession {
    __weak typeof(self) weakSelf = self;
    
    PLKLayerTokenConfiguration *configuration = [PLKLayerTokenConfiguration createWithToken:self.linkToken onSuccess:^(PLKLinkSuccess *success) {
        NSLog(@"public-token: %@ metadata: %@", success.publicToken, success.metadata.metadataJSON);
    }];
    
    configuration.onExit = ^(PLKLinkExit *exit) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) return;
        
        if (exit.error) {
            NSLog(@"exit with %@\n%@", exit.error, exit.metadata.metadataJSON);
            [strongSelf setState:LoadingStateError];
            NSString *displayMessage = exit.error.userInfo[kPLKExitErrorDisplayMessageKey];
            strongSelf.errorMessage = displayMessage ?: exit.error.localizedDescription ?: @"Exited with error";
            [strongSelf updateUI];
        } else {
            NSLog(@"exit with %@", exit.metadata.metadataJSON);
        }
    };
    
    configuration.onEvent = ^(PLKLinkEvent *event) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) return;
        
        NSLog(@"Link Event: %@", event.eventName);

        // Wait for the Layer ready event
        if (event.eventName.value == PLKEventNameValueLayerReady) {
            [strongSelf setState:LoadingStateReady];
            [strongSelf updateUI];
        } else if (event.eventName.value == PLKEventNameValueLayerNotAvailable) {
            [strongSelf setState:LoadingStateError];
            strongSelf.errorMessage = @"Layer not available";
            [strongSelf updateUI];
        }
    };
    
    NSError *error = nil;
    id<PLKPlaidLayerSession> session = [PLKPlaid createPlaidLayerSessionWithConfiguration:configuration error:&error];
    
    if (error) {
        self.state = LoadingStateError;
        self.errorMessage = error.localizedDescription;
        [self updateUI];
    } else {
        self.layerSession = session;
    }
}

- (void)submitButtonTapped {
    if (self.layerSession) {
        PLKSubmissionData *submissionData = [[PLKSubmissionData alloc] init];
        submissionData.phoneNumber = self.phoneNumberTextField.text;
        submissionData.dateOfBirth = self.dobTextField.text;
        
        [self.layerSession submit:submissionData];
    }
}

- (void)launchButtonTapped {
    if (self.layerSession) {
        __weak typeof(self) weakSelf = self;
        [self.layerSession openWithPresentationHandler:^(UIViewController *viewController) {
            __strong typeof(weakSelf) strongSelf = weakSelf;
            if (strongSelf) {
                [strongSelf presentViewController:viewController animated:YES completion:nil];
            }
        } dismissalHandler:^(UIViewController *viewController) {
            [viewController dismissViewControllerAnimated:YES completion:nil];
        }];
    }
}

- (void)updateUI {
    dispatch_async(dispatch_get_main_queue(), ^{
        // Submit button is enabled whenever we have a layer session
        BOOL hasSession = self.layerSession != nil;
        self.submitButton.enabled = hasSession;
        self.submitButton.backgroundColor = hasSession ? [UIColor systemBlueColor] : [UIColor systemGrayColor];
        
        switch (self.state) {
            case LoadingStateLoading:
                self.launchButton.enabled = NO;
                self.launchButton.backgroundColor = [UIColor systemGrayColor];
                [self.activityIndicator startAnimating];
                self.errorLabel.hidden = YES;
                break;
                
            case LoadingStateReady:
                self.launchButton.enabled = YES;
                self.launchButton.backgroundColor = [UIColor systemBlueColor];
                [self.activityIndicator stopAnimating];
                self.errorLabel.hidden = YES;
                break;
                
            case LoadingStateError:
                self.launchButton.enabled = NO;
                self.launchButton.backgroundColor = [UIColor systemGrayColor];
                [self.activityIndicator stopAnimating];
                self.errorLabel.text = self.errorMessage;
                self.errorLabel.hidden = NO;
                break;
        }
    });
}

@end

File: LinkDemo-UIKit/LinkDemo-UIKit/ViewControllers/PlaidLayerSessionViewController.swift
//
//  PlaidLayerSessionViewController.swift
//  LinkDemo-UIKit
//
//  Copyright © 2026 Plaid Inc. All rights reserved.
//

import LinkKit
import UIKit

struct LayerSubmissionData: SubmissionData {
    var phoneNumber: String?
    var dateOfBirth: String?
    var params: [String: String]?
}

class PlaidLayerSessionViewController: UIViewController {

    init(linkToken: String) {
        self.linkToken = linkToken
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        title = "PlaidLayerSession"
        view.backgroundColor = .systemBackground

        setupUI()
        updateUI()
        createSession()
    }

    // MARK: - Link Setup

    private func createSession() {
        let configuration = LayerTokenConfiguration(
            token: linkToken,
            onSuccess: { linkSuccess in
                print("public-token: \(linkSuccess.publicToken) metadata: \(linkSuccess.metadata)")
            },
            onExit: { [weak self] linkExit in
                if let error = linkExit.error {
                    print("exit with \(error)\n\(linkExit.metadata)")
                    DispatchQueue.main.async {
                        self?.state = .error(error.displayMessage ?? "Exited with error")
                    }
                } else {
                    print("exit with \(linkExit.metadata)")
                }
            },
            onEvent: { [weak self] linkEvent in
                print("Link Event: \(linkEvent)")

                if linkEvent.eventName == .layerReady {
                    DispatchQueue.main.async {
                        self?.state = .ready
                    }
                } else if linkEvent.eventName == .layerNotAvailable {
                    DispatchQueue.main.async {
                        self?.state = .error("Layer not available")
                    }
                }
            }
        )

        do {
            let session = try Plaid.createPlaidLayerSession(configuration: configuration)
            self.layerSession = session
            updateUI()
        } catch {
            self.state = .error(error.localizedDescription)
        }
    }

    // MARK: - Submit User Data

    @objc private func submitButtonTapped() {
        guard let layerSession = layerSession else { return }

        let submissionData = LayerSubmissionData(
            phoneNumber: phoneTextField.text,
            dateOfBirth: dobTextField.text
        )
        layerSession.submit(data: submissionData)
    }

    // MARK: - Launch Layer

    @objc private func launchButtonTapped() {
        guard let layerSession = layerSession else { return }
        layerSession.open(using: .viewController(self))
    }

    // MARK: - UI

    private let linkToken: String
    private var layerSession: PlaidLayerSession?
    private var state: LoadingState = .loading {
        didSet {
            updateUI()
        }
    }

    private let titleLabel = UILabel()
    private let versionLabel = UILabel()
    private let errorLabel = UILabel()
    private let phoneTextField = UITextField()
    private let dobTextField = UITextField()
    private let submitButton = UIButton(type: .system)
    private let launchButton = UIButton(type: .system)
    private let activityIndicator = UIActivityIndicatorView(style: .medium)

    private func updateUI() {
        let hasSession = layerSession != nil

        switch state {
        case .loading:
            activityIndicator.startAnimating()
            launchButton.isEnabled = false
            launchButton.backgroundColor = .systemGray
            submitButton.isEnabled = hasSession
            submitButton.backgroundColor = hasSession ? .systemBlue : .systemGray
            errorLabel.isHidden = true
        case .ready:
            activityIndicator.stopAnimating()
            launchButton.isEnabled = true
            launchButton.backgroundColor = .systemBlue
            submitButton.isEnabled = hasSession
            submitButton.backgroundColor = hasSession ? .systemBlue : .systemGray
            errorLabel.isHidden = true
        case .error(let message):
            activityIndicator.stopAnimating()
            launchButton.isEnabled = false
            launchButton.backgroundColor = .systemGray
            submitButton.isEnabled = false
            submitButton.backgroundColor = .systemGray
            errorLabel.text = message
            errorLabel.isHidden = false
        }
    }

    private func setupUI() {
        let scrollView = UIScrollView()
        scrollView.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(scrollView)

        let stackView = UIStackView()
        stackView.axis = .vertical
        stackView.spacing = 16
        stackView.translatesAutoresizingMaskIntoConstraints = false
        scrollView.addSubview(stackView)

        titleLabel.text = "Plaid Layer Session Example"
        titleLabel.font = .boldSystemFont(ofSize: 20)
        titleLabel.textAlignment = .center
        titleLabel.numberOfLines = 0
        stackView.addArrangedSubview(titleLabel)

        versionLabel.text = "LinkKit"
        versionLabel.font = .systemFont(ofSize: 14)
        versionLabel.textColor = .secondaryLabel
        versionLabel.textAlignment = .center
        stackView.addArrangedSubview(versionLabel)

        errorLabel.font = .systemFont(ofSize: 14)
        errorLabel.textColor = .systemRed
        errorLabel.textAlignment = .center
        errorLabel.numberOfLines = 0
        errorLabel.isHidden = true
        stackView.addArrangedSubview(errorLabel)

        let phoneLabel = UILabel()
        phoneLabel.text = "User Phone Number"
        phoneLabel.font = .systemFont(ofSize: 14)
        phoneLabel.textColor = .secondaryLabel
        stackView.addArrangedSubview(phoneLabel)

        phoneTextField.placeholder = "+1 415-555-0011"
        phoneTextField.text = "+1 415-555-0011"
        phoneTextField.keyboardType = .phonePad
        phoneTextField.borderStyle = .roundedRect
        stackView.addArrangedSubview(phoneTextField)

        let dobLabel = UILabel()
        dobLabel.text = "User Date of Birth"
        dobLabel.font = .systemFont(ofSize: 14)
        dobLabel.textColor = .secondaryLabel
        stackView.addArrangedSubview(dobLabel)

        dobTextField.placeholder = "YYYY-MM-DD"
        dobTextField.text = "1975-01-18"
        dobTextField.keyboardType = .numbersAndPunctuation
        dobTextField.borderStyle = .roundedRect
        stackView.addArrangedSubview(dobTextField)

        submitButton.setTitle("Submit User Data", for: .normal)
        submitButton.titleLabel?.font = .boldSystemFont(ofSize: 16)
        submitButton.setTitleColor(.white, for: .normal)
        submitButton.backgroundColor = .systemBlue
        submitButton.layer.cornerRadius = 10
        submitButton.addTarget(self, action: #selector(submitButtonTapped), for: .touchUpInside)
        submitButton.translatesAutoresizingMaskIntoConstraints = false
        stackView.addArrangedSubview(submitButton)

        launchButton.setTitle("Launch Layer", for: .normal)
        launchButton.titleLabel?.font = .boldSystemFont(ofSize: 16)
        launchButton.setTitleColor(.white, for: .normal)
        launchButton.backgroundColor = .systemBlue
        launchButton.layer.cornerRadius = 10
        launchButton.addTarget(self, action: #selector(launchButtonTapped), for: .touchUpInside)
        launchButton.translatesAutoresizingMaskIntoConstraints = false
        stackView.addArrangedSubview(launchButton)

        activityIndicator.color = .white
        activityIndicator.hidesWhenStopped = true
        activityIndicator.translatesAutoresizingMaskIntoConstraints = false
        launchButton.addSubview(activityIndicator)

        NSLayoutConstraint.activate([
            scrollView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            scrollView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            scrollView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            scrollView.bottomAnchor.constraint(equalTo: view.bottomAnchor),

            stackView.topAnchor.constraint(equalTo: scrollView.topAnchor, constant: 16),
            stackView.leadingAnchor.constraint(equalTo: scrollView.leadingAnchor, constant: 16),
            stackView.trailingAnchor.constraint(equalTo: scrollView.trailingAnchor, constant: -16),
            stackView.bottomAnchor.constraint(equalTo: scrollView.bottomAnchor, constant: -16),
            stackView.widthAnchor.constraint(equalTo: scrollView.widthAnchor, constant: -32),

            submitButton.heightAnchor.constraint(equalToConstant: 56),
            launchButton.heightAnchor.constraint(equalToConstant: 56),

            activityIndicator.leadingAnchor.constraint(equalTo: launchButton.leadingAnchor, constant: 16),
            activityIndicator.centerYAnchor.constraint(equalTo: launchButton.centerYAnchor),
        ])
    }
}
